examples/marching/src/host/paths.ads

```diff
@@ -0,0 +1,60 @@
+with Ada.Command_Line;
+with GNAT.OS_Lib;
+with Ada.Strings; use Ada.Strings;
+with Ada.Strings.Fixed; use Ada.Strings.Fixed;
+with Ada.Strings.Unbounded;
+
+package Paths is
+    type Path is new String;
+
+    function "+" (P : Path)
+      return String
+    is (String (P));
+
+    function "+" (S : Ada.Strings.Unbounded.Unbounded_String)
+      return Path
+    is (Path (Ada.Strings.Unbounded.To_String (S)));
+
+    function Is_Absolute (P : Path)
+      return Boolean is (GNAT.OS_Lib.Is_Absolute_Path (+P));
+
+    function Is_Directory (P : Path)
+      return Boolean is (GNAT.OS_Lib.Is_Directory (+P));
+
+    function Is_Regular_File (P : Path)
+      return Boolean is (GNAT.OS_Lib.Is_Regular_File (+P));
+
+    function Exists (P : Path)
+      return Boolean is (Is_Regular_File (P) or Is_Directory (P));
+
+    function "/" (P1 : Path; P2 : Path)
+      return Path
+    is (P1 & GNAT.OS_Lib.Directory_Separator & P2)
+      with Pre => not Is_Absolute (P2);
+
+    Directory_Separator_Str : constant String (1 .. 1) := (1 => GNAT.OS_Lib.Directory_Separator);
+
+    function Parent (P : Path)
+      return Path
+    is (P
+         (P'First
+          .. Index (+P, Directory_Separator_Str, Going=>Backward) - 1))
+      with Pre => Index (+P, Directory_Separator_Str) /= 0 -- no path separator, no parent
+                  and P'Length > 1; --  single path separator, no parent
+
+    function Root_Dir
+      return Path
+      --  Root dir of the project, based on the fact that main is placed in it
+      --  !!! Bug if main is in PATH and called without any path
+    is (Parent
+         (Path
+           (GNAT.OS_Lib.Normalize_Pathname (Ada.Command_Line.Command_Name))))
+      with Post => Is_Absolute (Root_Dir'Result)
```

Comment:
AFAIU Is_Absolute protects against the bug you're describing in the comment - would it be possible to also add a message explaining what went wrong to the user?

Reply:
No, this would require resolving the name of the command by looking into PATH and more (aliases...)

eg if I'm setting my PATH to `/toto/bin` which contains main I can call
```
$ main
```

In which case the Command_Name will simply be `main`, at this point I have no way to get `/toto/bin`, and this directory is not CWD, so no context apart from the POSIX PATH (and shell-specifics such as aliases) can tell me where is the command.

Reply:
I replaced the command by a Prem hope this clears it up

Reply:
Right, I understand why we can't figure out where the shader is - I was wondering if we could tell the user "hey, we expected a shader to exist in a path relative to the binary and we can't find it".

Reply:
I see, we're building without the contracts, so there's actually a shader loading error with the path of the shader that was tried (CWD) in that case.

If we want to do that with contracts (why would we though?) the default message will be that the command_name does not have a path separator, which may seem weird but is the true culprit (no path sep = path from CWD). Doing otherwise would require redispatching that exception from ui.adb, which is kind of ugly